tests/test_diff.py

```diff
@@ -0,0 +1,118 @@
+    co1.close()
+
+    repo.create_branch('testbranch')
+    co2 = repo.checkout(write=True, branch_name='testbranch')
+    for idx in range(10, 20):
+        dummyData[:] = idx
+        co2.datasets['dummy'][str(idx)] = dummyData
+    co2.metadata['foo'] = 'bar'
+    co2.commit('first commit on test branch adding non-conflict data and meta')
+    co2.close()
+    return repo
+
+
+@pytest.fixture()
+def repo_2_br_no_conf(repo_1_br_no_conf):
+    dummyData = np.arange(50)
+    repo = repo_1_br_no_conf
+    co1 = repo.checkout(write=True, branch_name='master')
+    for idx in range(20, 30):
+        dummyData[:] = idx
+        co1.datasets['dummy'][str(idx)] = dummyData
+    co1.commit('second commit on master adding non-conflict data')
+    co1.close()
+    return repo
+
+
+@pytest.fixture()
+def repo_2_br_with_conf(repo_2_br_no_conf):
+    dummyData = np.arange(50)
+    repo = repo_2_br_no_conf
+    dummyData[:] = 1234
+    co = repo.checkout(write=True, branch_name='master')
+    co.datasets['dummy']['15'] = dummyData
+    co.commit('final commit on master, making a conflict')
+    co.close()
+    return repo
+
+
+class TestReaderDiff:
+
+    def test_diff_by_commit_and_branch(self, repo_2_br_no_conf):
+        repo = repo_2_br_no_conf
+        testco = repo.checkout(branch_name='testbranch')
+        masterco = repo.checkout(branch_name='master')
+        commit_diffs = masterco.diff.commit(testco.commit_hash)
+        branch_diffs = masterco.diff.branch('testbranch')
+        assert commit_diffs == branch_diffs
+
+    def test_diff_with_wrong_commit_hash(self, repo_2_br_no_conf):
+        repo = repo_2_br_no_conf
+        testco = repo.checkout(branch_name='testbranch')
+        masterco = repo.checkout(branch_name='master')
+        wrong_commit_hash = testco.commit_hash + 'WrongHash'
+        with pytest.raises(ValueError):
+            masterco.diff.commit(wrong_commit_hash)
+
+    def test_diff_with_wrong_branch_name(self, repo_1_br_no_conf):
+        repo = repo_1_br_no_conf
+        masterco = repo.checkout(branch_name='master')
+        with pytest.raises(ValueError):
+            masterco.diff.branch('wrong_branch_name')
+
+    def test_diff_data_samples(self, repo_1_br_no_conf):
+        repo = repo_1_br_no_conf
+        dummyData = np.arange(50)
+
+        # mutating and removing data from testbranch
+        testco = repo.checkout(write=True, branch_name='testbranch')
+        testco.datasets['dummy']['1'] = dummyData
+        del testco.datasets['dummy']['2']
+        testco.commit("mutation and removal")
+        testco.close()
+
+        co = repo.checkout(branch_name='master')
+        diffdata = co.diff.branch('testbranch')
+        conflict_dict = diffdata[1]
+        assert conflict_dict['conflict_found'] is False
+        diffs1 = diffdata[0]
+
+        co = repo.checkout(branch_name='testbranch')
+        diffdata = co.diff.branch('master')
+        diffs2 = diffdata[0]
+
```

Comment:
@rlizzo I have a question in the above block. As per my understanding, Diff created at line#91 & line#97 should have same data but flipped between `master` and `dev` (btw: this terminology will be super confusing for us in the future. We should name it as source/destination or something IMO). Is that right?

Reply:
So you are correct, since this is a Fast-Forward merge (as shown below). However, it should generally hold true even for 3-way merges (I think)

![image](https://user-images.githubusercontent.com/17350115/59408181-68cbe000-8d81-11e9-8429-7068c2a07794.png)

However, there is a bug which would not display the information correctly from the diff (which is corrected in #77). Try rebasing this on top of master and let me know if that makes it easier to understand? 

After the fix this is was shows: 

![image](https://user-images.githubusercontent.com/17350115/59408280-b5172000-8d81-11e9-9075-7ef9159d9b45.png)

and 

![image](https://user-images.githubusercontent.com/17350115/59408313-c3fdd280-8d81-11e9-9554-130366c9d1e9.png)

Reply:
Gotcha! thanks

Reply:
Okey, thinking about it with a git analogy raises more questions. Perhaps we could talk about it over the call.

Reply:
absolutely!